File: 2925-maximum-score-after-applying-operations-on-a-tree/2925-maximum-score-after-applying-operations-on-a-tree.java
class Solution {
    public long maximumScoreAfterOperations(int[][] edges, int[] values) {
        
        int n = values.length;
        List<Integer>[] graph = new ArrayList[n];
        int[][] dp = new int[n][2];

        for (int i=0; i<n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for (int i=0; i<n-1; i++) {
            int from = edges[i][0];
            int to = edges[i][1];
            graph[from].add(to);
            graph[to].add(from);
        }
        
        long[] res = getSubTreeSum(0, -1, graph, values);

        return Math.max(res[0], res[1]);
    }
    
    private long[] getSubTreeSum(int i, int p, List<Integer>[] graph, int[] values) {
        if (i!=0 && graph[i].size()==1) {
            return new long[]{0, 0};
        }

        long[] res = new long[2];
        // remove root
        res[0] = 0;
        // not remove root
        res[1] += values[i];

        for (Integer child: graph[i]) {
            if ((int) child == p) {
                continue;
            }
            long[] childSum = getSubTreeSum(child, i, graph, values);
            res[0] += (childSum[0]+values[child]);
            res[1] += Math.max(childSum[0], childSum[1]);
        }

        return res;
    }
}

File: 2925-maximum-score-after-applying-operations-on-a-tree/2925-maximum-score-after-applying-operations-on-a-tree.cpp
class Solution {
public:
    // time/space: O(n)/O(n)
    long long maximumScoreAfterOperations(vector<vector<int>>& edges, vector<int>& values) {
        int n = values.size();
        // build the graph (tree)
        vector<vector<int>> graph(n);
        for (auto edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);    
        }
        
        vector<vector<long long>> dp(n, vector<long long>(2, -1LL));
        return helper(graph, values, dp, 0, -1, 0);
    }
private:
    long long helper(vector<vector<int>>& graph, vector<int>& values, vector<vector<long long>>& dp, int curr, int prev, int isHealthy) {
        // terminate
        if (dp[curr][isHealthy] != -1LL) return dp[curr][isHealthy];
        if ((graph[curr].size() == 1) && (graph[curr][0] == prev)) {
            // the leaf node must be healthy
            if (isHealthy == 0) return dp[curr][isHealthy] = 0LL;
            return dp[curr][isHealthy] = values[curr];
        }
        
        // enumerate
        long long take = values[curr];
        long long skip = 0;
        for (auto next : graph[curr]) {
            if (next == prev) continue;
            take += helper(graph, values, dp, next, curr, isHealthy);
            skip += helper(graph, values, dp, next, curr, 1);
        }
        return dp[curr][isHealthy] = max(take, skip); 
    }
};